### PR TITLE
PID-CPR Match

### PR DESCRIPTION
**Pull request by ludvigandersen**
Points to master now @davideluque :smile:

### CODE REVIEW

File: lib/nemid/pid_cpr.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
+require 'savon'
+
+module NemID
+    class PidCpr
+
+        def initialize(spid, cpr, pid, cert, key)
```
[the remark this answers is not in the document]
Comment (ludvigandersen) [reply lib/nemid/pid_cpr.rb:6]: I tried using the crypto module for extracting the key and certificate, i came upon some issues, don't remember exactly, but i'll try it out again and see what it was that occured.
The key and cert, at current times, have been manually extracted and i should definitely see how to just pass in the VOCES and let the gem extract the key and cert.

The PID/CPR doesn't find the CPR, it matches the CPR with the PID.
PID/CPR Lookup is what you are thinking about, and is only used by public organizations, while the PID/CPR match is used by private organizations.

[the remark this answers is not in the document]
Comment (ludvigandersen) [reply lib/nemid/pid_cpr.rb:6]: The user provides the CPR in the reset password flow, which is sent to the backend.

File: lib/nemid/pid_cpr.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
+require 'savon'
+
+module NemID
+    class PidCpr
```
[the remark this answers is not in the document]
Comment (ludvigandersen) [reply lib/nemid/pid_cpr.rb:4]: You are right.

**Comment by ludvigandersen** [comment]
> Can you point me to the document where the PIDCPR service is explained? Tyyyyy <3

https://www.nets.eu/dk-da/kundeservice/nemid-tjenesteudbyder/The-NemID-service-provider-package-of-JavaScript-(english%20version)/Documents/Specification%20document%20for%20the%20PID-CPR%20service.pdf

https://www.nets.eu/dk-da/kundeservice/nemid-tjenesteudbyder/NemID-tjenesteudbyderpakken/Documents/Configuration%20and%20setup.pdf

File: lib/nemid/pid_cpr.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    16 = BRUTE_FORCE_ATTEMPT_DETECTED ("Forsøg på systematisk søgning på CPR", "Brute force attempt detected") 
+    17 = NOT_AUTHORIZED_FOR_SERVICE_LOOKUP ( "Der er ikke rettighed til at foretage opslag på service", "Caller not authorized for service lookup") 
+    4096 = NOT_PID_SERVICE_REQUEST ("Modtaget message ikke pidCprRequest eller pidCprServerStatus", "Non request XML received") 
+    8192 = XML_PARSE_ERROR ("XML pakke kan ikke parses", "Non parsable XML received") 
+    8193 = XML_VERSION_NOT_SUPPORTED ("Version er ikke understøttet", "Version not supported") 
+    8194 = PID_DOES_NOT_MATCH_BASE64_CERTIFICATE ("PID stemmer med ikke med certifikat", "PID does not match certifikat") 
+    8195 = MISSING_CLIENT_CERT ("Klient certifikat ikke præsenteret", "No client certificate presented") 
+    16384 = INTERNAL_ERROR ("Intern DanID fejl", "Internal DanID error")
+=end
+        def pid_cpr_match
```
[the remark this answers is not in the document]
Comment (ludvigandersen) [reply lib/nemid/pid_cpr.rb:30]: I quite like that, and yes i agree that the initialization of the module gets quite bloated with all the arguments.

**Comment by ludvigandersen** [comment]
I am not fully confident in the way the cert and key are extracted, saved in a file and the deleted.
Not sure if it a security concern saving the files like that.

**Review by davideluque — APPROVED** [review]

**Comment by davideluque** [comment]
I think this feature will change the lives of many Danes @ludvigandersen 

They don't know what is coming.

**Pull request merged**